src/impl_/trampoline.rs

```diff
@@ -0,0 +1,173 @@
+
+    let gil = GILPool::new();
+    let py = gil.python();
+    panic_result_into_callback_output(py, std::panic::catch_unwind(move || f(py, slf)))
+}
+
+#[inline]
+pub unsafe fn setter(
+    slf: *mut ffi::PyObject,
+    value: *mut ffi::PyObject,
+    closure: *mut c_void,
+    f: for<'py> unsafe fn(Python<'py>, *mut ffi::PyObject, *mut ffi::PyObject) -> PyResult<c_int>,
+) -> c_int {
+    // PyO3 doesn't use the closure argument at present.
+    debug_assert!(closure.is_null());
+
+    let gil = GILPool::new();
+    let py = gil.python();
+    panic_result_into_callback_output(py, std::panic::catch_unwind(move || f(py, slf, value)))
+}
+
+// Trampolines used by slot methods
+trampolines!(
+    pub fn binaryfunc(slf: *mut ffi::PyObject, arg1: *mut ffi::PyObject) -> *mut ffi::PyObject;
+
+    pub fn descrgetfunc(
+        slf: *mut ffi::PyObject,
+        arg1: *mut ffi::PyObject,
+        arg2: *mut ffi::PyObject,
+    ) -> *mut ffi::PyObject;
+
+    pub fn getiterfunc(slf: *mut ffi::PyObject) -> *mut ffi::PyObject;
+
+    pub fn hashfunc(slf: *mut ffi::PyObject) -> ffi::Py_hash_t;
+
+    pub fn inquiry(slf: *mut ffi::PyObject) -> c_int;
+
+    pub fn iternextfunc(slf: *mut ffi::PyObject) -> *mut ffi::PyObject;
+
+    pub fn lenfunc(slf: *mut ffi::PyObject) -> ffi::Py_ssize_t;
+
+    pub fn newfunc(
+        subtype: *mut ffi::PyTypeObject,
+        args: *mut ffi::PyObject,
+        kwargs: *mut ffi::PyObject,
+    ) -> *mut ffi::PyObject;
+
+    pub fn objobjproc(slf: *mut ffi::PyObject, arg1: *mut ffi::PyObject) -> c_int;
+
+    pub fn reprfunc(slf: *mut ffi::PyObject) -> *mut ffi::PyObject;
+
+    pub fn richcmpfunc(
+        slf: *mut ffi::PyObject,
+        other: *mut ffi::PyObject,
+        op: c_int,
+    ) -> *mut ffi::PyObject;
+
+    pub fn ssizeargfunc(arg1: *mut ffi::PyObject, arg2: ffi::Py_ssize_t) -> *mut ffi::PyObject;
+
+    pub fn ternaryfunc(
+        slf: *mut ffi::PyObject,
+        arg1: *mut ffi::PyObject,
+        arg2: *mut ffi::PyObject,
+    ) -> *mut ffi::PyObject;
+
+    pub fn unaryfunc(slf: *mut ffi::PyObject) -> *mut ffi::PyObject;
+
+    pub fn dealloc(slf: *mut ffi::PyObject) -> ();
+);
+
+#[cfg(any(not(Py_LIMITED_API), Py_3_11))]
+trampolines! {
+    pub fn getbufferproc(slf: *mut ffi::PyObject, buf: *mut ffi::Py_buffer, flags: c_int) -> c_int;
+
+    pub fn releasebufferproc(slf: *mut ffi::PyObject, buf: *mut ffi::Py_buffer) -> ();
+}
+
+// Ipowfunc is a unique case where PyO3 has its own type
+// to workaround a problem on 3.7 (see IPowModulo type definition).
+// Once 3.7 support dropped can just remove this.
+trampoline!(
+    pub fn ipowfunc(
+        arg1: *mut ffi::PyObject,
+        arg2: *mut ffi::PyObject,
+        arg3: IPowModulo,
+    ) -> *mut ffi::PyObject;
+);
+
+#[inline]
+pub fn trampoline_inner<F, R>(body: F) -> R
+where
+    F: for<'py> FnOnce(Python<'py>) -> PyResult<R> + UnwindSafe,
+    R: PyCallbackOutput,
+{
+    let trap = PanicTrap::new("uncaught panic at ffi boundary");
```

Comment:
@adamreichold this is the solution I had in mind - now all the trampolines use a `PanicTrap` to guard against any panics leaking out of them.

Reply:
This does put the `PanicTrap` closer to the FFI boundary itself, which is nice.

But I think the `getter` and `setter` trampolines are not covered by this? Couldn't they use `trampoline_inner` as well?

Also, wouldn't `pub(crate)` suffice for this function as the only placing which calls this outside of `mod trampoline` is within `mod pyclass`?

Personally, I think I would have felt safer if the `PanicTrap` was installed inside `panic_result_into_callback_output`, but admittedly that function itself does not imply an FFI boundary. Are there any places we use this function outside of FFI interactions at the moment?

Reply:
Ah, yes corrected `getter` and `setter` and made it `pub(crate)`.

Regarding `panic_result_into_callback_output`, I somewhat felt similar but on reflection I decided that the safety issue is not the panic payload drop panicking, but that secondary panic then making it past the FFI boundary unprotected. And on that reflection it seemed clearer to me that the responsibility for `PanicTrap` is with these trampolines.

Reply:
> And on that reflection it seemed clearer to me that the responsibility for PanicTrap is with these trampolines.

Agreed. We might want to make another sweep over all users of `panic_result_into_callback_output` though whether those need similar treatment.

From a quick grep, it might make sense to investigate

* `ModuleDef::module_init` (Might be benign as no custom code is called, but why the `catch_unwind` then in the first place?)
* `macro_rules! $generate_macro`
* `macro_rules! generate_pyclass_getattro_slot`

Also, we might wanto adjust the module-level doc comment for `mod pycell` which discusses the generated code. (Or we might not want to do that since the trampolines are an optimization and would only complicate the exposition without adding anything to the discussion of aliasing and mutability aimed at over there.)

Reply:
Great observation. I've pushed a commit which makes `panic_result_into_callback_output` a private function in `trampoline.rs` and converted all callsites listed above to also make use of trampolines.

I did adjust the `mod pycell` documentation a bit, given I removed `panic_result_into_callback_output` from the (hidden) crate exports.

Reply:
This was certainly sweeping! I feel much better with `panic_result_into_callback_output` being a module-private function since almost everything related to crossing the FFI boundary is now here.